### PR TITLE
all: use a proper log id which can reference channels, subchannels, a…

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
…nd transports

### CODE REVIEW

File: core/src/main/java/io/grpc/internal/DelayedClientTransport.java [hunk 1]
```diff
@@ -367,9 +367,10 @@ void endBackoff() {
     }
   }
 
+  // TODO(carl-mastrangelo): remove this once the Subchannel change is in.
   @Override
-  public String getLogId() {
-    return GrpcUtil.getLogId(this);
+  public LogId getLogId() {
+    return LogId.allocate(getClass().getName());
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/DelayedClientTransport.java]: Even if this is temporary, we should not allocate a new ID every single time `getLogId()` is called.

File: core/src/main/java/io/grpc/internal/LogId.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A loggable ID, unique for the duration of the program.
+ */
+public final class LogId {
+  private static final AtomicLong idAlloc = new AtomicLong();
+
+  public static LogId allocate(String name) {
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/LogId.java]: Javadocs to explain what `name` is.

File: core/src/main/java/io/grpc/internal/LogId.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  private LogId(String source, long id) {
+    this.source = source;
+    this.id = id;
+  }
+
+  public long getId() {
+    return id;
+  }
+
+  public String getSource() {
```
Comment (zhangkun83) [inline core/src/main/java/io/grpc/internal/LogId.java]: You have different terminologies (source vs name) for the same thing. There should be only one.

**Comment by carl-mastrangelo** [comment]
PTAL

**Pull request merged**